artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/LinkedListImpl.java

```diff
@@ -46,13 +49,59 @@
 
    private final Comparator<E> comparator;
 
+   private ToLongFunction<E> idSupplier;
+
+   LongObjectHashMap<Node<E>> nodeMap;
+
    public LinkedListImpl() {
-      this(null);
+      this(null, null);
    }
 
+
    public LinkedListImpl(Comparator<E> comparator) {
+      this(comparator, null);
+   }
+
+   @Override
+   public void clearID() {
+      idSupplier = null;
+      nodeMap.clear(); // just a little hand to GC
+      nodeMap = null;
+   }
+
+   @Override
+   public void setIDSupplier(ToLongFunction<E> supplier) {
+      this.idSupplier = supplier;
+      nodeMap = new LongObjectHashMap<>(this.size);
+
+      Iterator iterator = (Iterator)iterator();
+      try {
+         while (iterator.hasNext()) {
+            E value = iterator.next();
+            Node<E> position = iterator.last;
+            putID(value, position);
+         }
+      } finally {
+         iterator.close();
+      }
+   }
+
+   private void putID(E value, Node<E> position) {
+      long id = idSupplier.applyAsLong(value);
+      if (id >= 0) {
+         nodeMap.put(id, position);
+      }
+   }
+
+   public LinkedListImpl(Comparator<E> comparator, ToLongFunction<E> supplier) {
       iters = createIteratorArray(INITIAL_ITERATOR_ARRAY_SIZE);
       this.comparator = comparator;
+      this.idSupplier = supplier;
+      if (idSupplier != null) {
+         this.nodeMap = new LongObjectHashMap<>(this.size);
```

Comment:
`this.size is 0` here

Reply:
It shouldn't be 0... we could eventually add the supplier when there are items on the list already.

Reply:
I see what Franz is saying, that this could pass a very small value for size, likely 0 in cases (e.g on empty sub lists of the priority list for example?), which will then force the map to grow on its first use and then several times quickly after (peeking, looks like it might do 1 entry to 2 on first use, then 2 to 4, 4 to 8, etc).

The default size looks to be 8 when none is given, might might be a reasonable minimum to pass? If space efficiency is a concern, perhaps the impl be altered to only create the nodeMap at all if it doesnt exist when wanting to insert something? (I'm assuming the oveall usage of the list is thread safe, otherwise it woud all be inherently broken already anyway)

Reply:
Looking at the implementation, the size will growth by 50th (0.5).

if I know how many elements I'm inserting it seems more efficient to set the size ahead instead of keeping growing the elements and moving them between the temp arrays and new arrays as the implementation does.

Reply:
We werent saying dont use the size at all, but rather dont use it when it is known to be very small, e.g 0 like it currently mould. Instead, it could use e.g max( size, <some-mininum>).

The load factor is 0.5, which I believe means it grows and rehashes when exceeding 50% capacity. It doubles the size when it grows as it uses powers of 2 for the sizing.

Reply:
OIC.. it makes sense! thanks

Reply:
although the implementation is checking it... but I will pass in Math.min (8, size)

Reply:
Yep it works as is, its just a bit silly creating a map with a single element array that will resize on first use (unless expecting never to have more than 0/1 elements, in which case why use a map).